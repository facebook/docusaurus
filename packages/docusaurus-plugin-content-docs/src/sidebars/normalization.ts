/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import _ from 'lodash';
import logger from '@docusaurus/logger';
import {isCategoriesShorthand} from './utils';
import type {
  NormalizedSidebarItem,
  NormalizedSidebar,
  NormalizedSidebars,
  SidebarCategoriesShorthand,
  SidebarItemCategoryConfig,
  SidebarItemConfig,
  SidebarConfig,
  SidebarsConfig,
  NormalizedSidebarItemCategory,
} from './types';

function normalizeCategoriesShorthand(
  sidebar: SidebarCategoriesShorthand,
): SidebarItemCategoryConfig[] {
  return Object.entries(sidebar).map(([label, items]) => ({
    type: 'category',
    label,
    items,
  }));
}

/**
 * Normalizes recursively item and all its children. Ensures that at the end
 * each item will be an object with the corresponding type.
 */
export function normalizeItem(
  item: SidebarItemConfig,
): NormalizedSidebarItem[] {
  if (typeof item === 'string') {
    return [{type: 'doc', id: item}];
  }
  if (isCategoriesShorthand(item)) {
    // This will never throw anyways
    return normalizeSidebar(item, 'sidebar items slice');
  }
  if (
    (item.type === 'doc' || item.type === 'ref') &&
    typeof item.label === 'string'
  ) {
    return [{...item, translatable: true}];
  }
  if (item.type === 'category') {
    const normalizedCategory: NormalizedSidebarItemCategory = {
      ...item,
      items: normalizeSidebar(
        item.items,
        logger.interpolate`code=${'items'} of the category name=${item.label}`,
      ),
    };
    return [normalizedCategory];
  }
  return [item];
}

function normalizeSidebar(
  sidebar: SidebarConfig,
  place: string,
): NormalizedSidebar {
  if (!Array.isArray(sidebar) && !isCategoriesShorthand(sidebar)) {
    throw new Error(
      logger.interpolate`Invalid sidebar items collection code=${JSON.stringify(
        sidebar,
      )} in ${place}: it must either be an array of sidebar items or a shorthand notation (which doesn't contain a code=${'type'} property). See url=${'https://docusaurus.io/docs/sidebar/items'} for all valid syntaxes.`,
    );
  }

  const normalizedSidebar = Array.isArray(sidebar)
    ? sidebar
    : normalizeCategoriesShorthand(sidebar);

  return normalizedSidebar.flatMap((subItem) => normalizeItem(subItem));
}

export function normalizeSidebars(
  sidebars: SidebarsConfig,
): NormalizedSidebars {
  return _.mapValues(sidebars, (sidebar, id) =>
    normalizeSidebar(sidebar, logger.interpolate`sidebar name=${id}`),
  );
}
