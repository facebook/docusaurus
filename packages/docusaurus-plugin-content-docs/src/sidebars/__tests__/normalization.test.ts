/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {normalizeSidebars} from '../normalization';

describe('normalization', () => {
  it('normalizes shorthands', () => {
    expect(
      normalizeSidebars({
        sidebar: {
          Category: ['doc1', 'doc2'],
          'Category 2': {
            'Subcategory 1': ['doc3', 'doc4'],
            'Subcategory 2': ['doc5', 'doc6'],
          },
        },
      }),
    ).toMatchSnapshot();

    expect(
      normalizeSidebars({
        sidebar: [
          {
            type: 'link',
            label: 'Google',
            href: 'https://google.com',
          },
          {
            'Category 1': ['doc1', 'doc2'],
            'Category 2': ['doc3', 'doc4'],
          },
        ],
      }),
    ).toMatchSnapshot();
  });
  it('rejects some invalid cases', () => {
    expect(() =>
      normalizeSidebars({
        sidebar: {
          // @ts-expect-error: test
          Category: {type: 'autogenerated', dirName: 'foo'},
          // @ts-expect-error: test
          Category2: {type: 'autogenerated', dirName: 'bar'},
        },
      }),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Invalid sidebar items collection \`{"type":"autogenerated","dirName":"foo"}\` in \`items\` of the category Category: it must either be an array of sidebar items or a shorthand notation (which doesn't contain a \`type\` property). See https://docusaurus.io/docs/sidebar/items for all valid syntaxes."`,
    );

    expect(() =>
      normalizeSidebars({
        sidebar: [
          'foo',
          {
            Category: {
              // @ts-expect-error: test
              type: 'category',
              items: ['bar', 'baz'],
            },
          },
        ],
      }),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Invalid sidebar items collection \`{"type":"category","items":["bar","baz"]}\` in \`items\` of the category Category: it must either be an array of sidebar items or a shorthand notation (which doesn't contain a \`type\` property). See https://docusaurus.io/docs/sidebar/items for all valid syntaxes."`,
    );

    expect(() =>
      normalizeSidebars({
        sidebar: [
          'foo',
          {
            // @ts-expect-error: test
            Category: 'bar',
          },
        ],
      }),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Invalid sidebar items collection \`"bar"\` in \`items\` of the category Category: it must either be an array of sidebar items or a shorthand notation (which doesn't contain a \`type\` property). See https://docusaurus.io/docs/sidebar/items for all valid syntaxes."`,
    );

    expect(() =>
      normalizeSidebars({
        // @ts-expect-error: test
        sidebar: 'item',
      }),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Invalid sidebar items collection \`"item"\` in sidebar sidebar: it must either be an array of sidebar items or a shorthand notation (which doesn't contain a \`type\` property). See https://docusaurus.io/docs/sidebar/items for all valid syntaxes."`,
    );
  });

  it('adds a translatable marker for labels defined in sidebars.js', () => {
    expect(
      normalizeSidebars({
        sidebar: [
          {
            type: 'doc',
            id: 'google',
            label: 'Google',
          },
          {
            'Category 1': ['doc1', 'doc2'],
            'Category 2': [
              'doc3',
              'doc4',
              {
                type: 'ref',
                id: 'msft',
                label: 'Microsoft',
              },
            ],
          },
        ],
      }),
    ).toMatchSnapshot();
  });
});
