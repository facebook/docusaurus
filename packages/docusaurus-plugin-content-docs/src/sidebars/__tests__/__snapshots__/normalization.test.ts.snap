// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalization adds a translatable marker for labels defined in sidebars.js 1`] = `
{
  "sidebar": [
    {
      "id": "google",
      "label": "Google",
      "translatable": true,
      "type": "doc",
    },
    {
      "items": [
        {
          "id": "doc1",
          "type": "doc",
        },
        {
          "id": "doc2",
          "type": "doc",
        },
      ],
      "label": "Category 1",
      "type": "category",
    },
    {
      "items": [
        {
          "id": "doc3",
          "type": "doc",
        },
        {
          "id": "doc4",
          "type": "doc",
        },
        {
          "id": "msft",
          "label": "Microsoft",
          "translatable": true,
          "type": "ref",
        },
      ],
      "label": "Category 2",
      "type": "category",
    },
  ],
}
`;

exports[`normalization normalizes shorthands 1`] = `
{
  "sidebar": [
    {
      "items": [
        {
          "id": "doc1",
          "type": "doc",
        },
        {
          "id": "doc2",
          "type": "doc",
        },
      ],
      "label": "Category",
      "type": "category",
    },
    {
      "items": [
        {
          "items": [
            {
              "id": "doc3",
              "type": "doc",
            },
            {
              "id": "doc4",
              "type": "doc",
            },
          ],
          "label": "Subcategory 1",
          "type": "category",
        },
        {
          "items": [
            {
              "id": "doc5",
              "type": "doc",
            },
            {
              "id": "doc6",
              "type": "doc",
            },
          ],
          "label": "Subcategory 2",
          "type": "category",
        },
      ],
      "label": "Category 2",
      "type": "category",
    },
  ],
}
`;

exports[`normalization normalizes shorthands 2`] = `
{
  "sidebar": [
    {
      "href": "https://google.com",
      "label": "Google",
      "type": "link",
    },
    {
      "items": [
        {
          "id": "doc1",
          "type": "doc",
        },
        {
          "id": "doc2",
          "type": "doc",
        },
      ],
      "label": "Category 1",
      "type": "category",
    },
    {
      "items": [
        {
          "id": "doc3",
          "type": "doc",
        },
        {
          "id": "doc4",
          "type": "doc",
        },
      ],
      "label": "Category 2",
      "type": "category",
    },
  ],
}
`;
